pkg/controller/pinned-image-set/pinned_image_set_controller.go

```diff
@@ -0,0 +1,698 @@
+		}
+		return ctrl.patch(imageSet.Name, patch)
+	})
+}
+
+func (ctrl *Controller) popFinalizer(imageSet *mcfgv1.PinnedImageSet) error {
+	return retry.RetryOnConflict(ctrl.waitBackOff, func() error {
+		newimageSet, err := ctrl.imageSetLister.Get(imageSet.Name)
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		curJSON, err := json.Marshal(newimageSet)
+		if err != nil {
+			return err
+		}
+
+		imageSetTmp := newimageSet.DeepCopy()
+		imageSetTmp.Finalizers = append(imageSet.Finalizers[:0], imageSet.Finalizers[1:]...)
+
+		modJSON, err := json.Marshal(imageSetTmp)
+		if err != nil {
+			return err
+		}
+
+		patch, err := jsonmergepatch.CreateThreeWayJSONMergePatch(curJSON, modJSON, curJSON)
+		if err != nil {
+			return err
+		}
+		return ctrl.patch(imageSet.Name, patch)
+	})
+}
+
+func (ctrl *Controller) patch(name string, patch []byte) error {
+	_, err := ctrl.mcfgClient.MachineconfigurationV1().PinnedImageSets().Patch(context.TODO(), name, types.MergePatchType, patch, metav1.PatchOptions{})
+	return err
+}
+
+func createTOMLgeneratedConfigFile(path string, tomlConf interface{}) (generatedConfigFile, error) {
+	var newData bytes.Buffer
+	encoder := toml.NewEncoder(&newData)
+	if err := encoder.Encode(tomlConf); err != nil {
+		return generatedConfigFile{}, fmt.Errorf("error encoding toml for CRIO drop-in files: %w", err)
+	}
+	return generatedConfigFile{filePath: path, data: newData.Bytes()}, nil
+}
+
+type tomlConfigCRIOPinnedImages struct {
+	Crio struct {
+		Image struct {
+			PinnedImages []mcfgv1.PinnedImageRef `toml:"pinned_images,omitempty"`
+		} `toml:"image"`
+	} `toml:"crio"`
+}
+
+// generatedConfigFile is a struct that holds the filepath and data of the various configs
+// Using a struct array ensures that the order of the ignition files always stay the same
+// ensuring that double MCs are not created due to a change in the order
+type generatedConfigFile struct {
+	filePath string
+	data     []byte
+}
+
+// getMachineConfigKey returns the managed key for the machine config
+func getMachineConfigKey(pool *mcfgv1.MachineConfigPool, client mcfgclientset.Interface, imageSetOrig *mcfgv1.PinnedImageSet) (string, error) {
+	imageSetListAll, err := client.MachineconfigurationV1().PinnedImageSets().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to list pinned image sets %v", err)
+	}
+
+	if imageSetListAll == nil || len(imageSetListAll.Items) == 0 {
+		return defaultManagedKey(pool, client)
+	}
+
+	var imageSetList []mcfgv1.PinnedImageSet
+	for _, imageSet := range imageSetListAll.Items {
+		selector, err := metav1.LabelSelectorAsSelector(imageSet.Spec.MachineConfigPoolSelector)
+		if err != nil {
+			return "", fmt.Errorf("invalid label selector: %w", err)
+		}
+		// if no mcp selector is defined then the image set is skipped
+		if selector.Empty() || !selector.Matches(labels.Set(pool.Labels)) {
+			// not associated with the desired pool
```

Comment:
I think some logging here would probably be helpful for debugging issues with this controller

Reply:
adding test as well